consensus/e2e_tests/state_sync_test.go

```diff
@@ -289,24 +286,43 @@ func TestStateSync_UnsynchedPeerSynchs_Success(t *testing.T) {
 	_, err = WaitForNetworkConsensusEvents(t, clockMock, eventsChannel, consensus.Prepare, consensus.Propose, numExpectedMsgs, 250, true)
 	require.NoError(t, err)
 
+	advanceTime(t, clockMock, 10*time.Millisecond)
 	for nodeId, pocketNode := range pocketNodes {
 		nodeState := GetConsensusNodeState(pocketNode)
-
+		fmt.Printf("Node state, node id: %d, height: %d, round: %d, step: %d, leaderId: %d \n", nodeId, nodeState.Height, nodeState.Round, nodeState.Step, nodeState.LeaderId)
 		assertNodeConsensusView(t, nodeId,
 			typesCons.ConsensusNodeState{
 				Height: testHeight,
 				Step:   uint8(consensus.Prepare),
-				Round:  uint8(testRound + 1),
+				Round:  uint8(currentRound),
 			},
 			nodeState)
+		require.Equal(t, leaderId, nodeState.LeaderId)
```

Comment:
The issue lies in `electNextLeaderDeterministicRoundRobin`. I don't know why, but if you add the following:

```go
	fmt.Println("OLSH DEBUG", value, numVals, int64(message.Height), int64(message.Round), int64(message.Step))
```

You'll see that `message.Height` is `2` instead of `3` when the failure happens. Try to look into how this happens and hopefully you'll uncover the source of the bug.

----

Also, I noticed that we don't lock in these two places. I'm always averse to locking, so I'm not saying we need to, but asking you to look into it in the context of state sync.

```bash
--- a/consensus/pacemaker_consensus.go
+++ b/consensus/pacemaker_consensus.go
@@ -61,6 +61,9 @@ func (m *consensusModule) IsLeaderSet() bool {
 }

 func (m *consensusModule) NewLeader(msg *anypb.Any) error {
+       m.m.Lock()
+       defer m.m.Unlock()
+
```

```bash
+++ b/consensus/debugging.go
@@ -47,6 +47,9 @@ func (m *consensusModule) GetAggregatedStateSyncMetadataMaxHeight() (maxHeight u
 }

 func (m *consensusModule) GetLeaderElectionResult(height, round uint64, step uint8) uint64 {
+       m.m.Lock()
+       defer m.m.Unlock()
+
```

Reply:
> I think the best thing to do now is to remove this test from this PR, and rather have it in #352 where the block actually receives blocks and stores. There, we must ensure leader election is performed upon syncing up to the latest state. Wdyt?

I would disagree.

This PR is almost 3 weeks of work and `TestStateSync_UnsynchedPeerSynchs_Success` is the only unit test we added. Let me know if I'm incorrect, but if it's the case, removing it is not an option.

Whenever only 1 (or few) unit tests are written, and they're difficult to write, incorrect or are flaky, it's a signal that the underlying work needs to be improved.